End event is called more then once

```sh
flow-app.js:1969 Error: no writecb in Transform class(…)
EventEmitter.emit @ flow-app.js:1969
EventEmitter.emit @ flow-app.js:1984
EventEmitter.emit @ flow-app.js:1984
afterTransform @ flow-app.js:4288
TransformState.afterTransform @ flow-app.js:4272
runSeq @ flow-app.js:2726
(anonymous function) @ view.js:1403
createTemplate @ view.js:1336
(anonymous function) @ view.js:1297
(anonymous function) @ flow-app.js:2278
EventEmitter.emit @ flow-app.js:1984
end @ flow-app.js:2451
EventEmitter.emit @ flow-app.js:1984
endReadableNT @ flow-app.js:4196
afterTick @ flow-app.js:3116
Item.run @ flow-app.js:3187
drainQueue @ flow-app.js:3157
```

--- a/lib/node.js
+++ b/lib/node.js
@@ -7,16 +7,32 @@ module.exports = function (scope, key) {
     // TODO make objectMode configurable
     var node = new EventEmitter();
     node._i = new PassThrough({objectMode: true});
-    node._i.on('end', end.bind(scope, node, '(from input)'));
     node._i.on('error', node.emit.bind(node, 'error'));
     node._i.cork();
 
     node.write = node._i.write.bind(node._i);
-    node.end = function (lastChunk) {
-        //end.call(scope, key, 'from method');
-        //node.ended = 1;
-        node._i.end(lastChunk);
-    };
+    node.end = function (chunk) {
+		delete scope.streams[node.name];
+
+        if (node.ended) {
+            return;
+        }
+
+		// wait until node is ready, so no bufferted chunks are missed
+        if (!node.ready) {
+            node.once('ready', node._i.end.bind(node._i, chunk));
+        } else {
+		    node._i.end(chunk);
+        }
+	};
+
+    node.on('end', function () {
+        //console.log('Flow.node.end(event):', node.name, node.ready, node.ended);
+        delete scope.streams[node.name];
+    });
+
+	// TODO find a better solution for substring that are closed (duplex, writable)
+    node.on('unpipe', node.end.bind(node));
 
     node.pipe = pipe;
     node.unpipe = node._i.unpipe.bind(node._i);
@@ -40,14 +56,15 @@ function pipe (dest) {
 }
 
 function end (node, origin) {
-    console.log('Flow.node.end:', node.name, origin, node.ready, node.ended);
 
-    if (!node.ended && node.ready) {
-        node.emit('end');
+    if (node.ended) {
+		return;
     }
-    node.ended = 1;
 
-    delete this.streams[node.name];
+    //console.log('Flow.node.end:', node.name, origin, node.ready, node.ended);
+
+    node.ended = 1;
+	node.emit('end');
 }
 
 function join (output) {
@@ -65,7 +82,7 @@ function join (output) {
 
         // pipe and set node output if ouput is duplex
         if ((output.readable && output.writable) || (output._i && output.join && output.link)) {
-//console.log('Duplex pipe to', (this._o ? '"output"' : '"input"'), 'and set as output.');
+//console.log((output.name || 'Seq of ' + this.name), 'Duplex pipe to', (this._o ? '"output"' : '"input"'), 'and set as output.');
             this._o = this._o ? this._o.pipe(output) : this.pipe(output);
 
         // overwrite output if sequence is readonly
@@ -145,9 +162,8 @@ function link (options, event) {
     }
 
     output.on('data', this.emit.bind(this, 'data'));
-    output.on('end', end.bind(node.scope, node, '(from output)'));
+	output.on('end', end.bind(node.scope, node, '(from output)'));
+    this._i.uncork();
 
 //console.log('Uncork input:', node.name, '..output exists?', (this._o ? 'Yes' : 'No'));
-
-    this._i.uncork();
 }

--- a/lib/node.js
+++ b/lib/node.js
@@ -40,11 +40,12 @@ function pipe (dest) {
 }
 
 function end (node, origin) {
-    console.log('Flow.node.end:', node.name, origin, node.ready);
+    console.log('Flow.node.end:', node.name, origin, node.ready, node.ended);
 
-    if (node.ready) {
+    if (!node.ended && node.ready) {
         node.emit('end');
     }
+    node.ended = 1;
 
     delete this.streams[node.name];
 }

--- a/lib/node.js
+++ b/lib/node.js
@@ -7,12 +7,14 @@ module.exports = function (scope, key) {
     // TODO make objectMode configurable
     var node = new EventEmitter();
     node._i = new PassThrough({objectMode: true});
+    node._i.on('end', end.bind(scope, node, '(from input)'));
     node._i.on('error', node.emit.bind(node, 'error'));
     node._i.cork();
 
     node.write = node._i.write.bind(node._i);
     node.end = function (lastChunk) {
-        end.call(scope, key);
+        //end.call(scope, key, 'from method');
+        //node.ended = 1;
         node._i.end(lastChunk);
     };
 
@@ -22,12 +24,7 @@ module.exports = function (scope, key) {
     node.join = join;
     node.link = link;
     node.name = key;
-
-    // end node on error
-    node.on('error', node.emit.bind(node, 'end'));
-
-    // end input and remove from cache
-    node.on('end', end.bind(scope, key));
+    node.scope = scope;
 
     return node;
 };
@@ -42,8 +39,14 @@ function pipe (dest) {
     return dest;
 }
 
-function end (key) {
-    delete this.streams[key];
+function end (node, origin) {
+    console.log('Flow.node.end:', node.name, origin, node.ready);
+
+    if (node.ready) {
+        node.emit('end');
+    }
+
+    delete this.streams[node.name];
 }
 
 function join (output) {
@@ -113,9 +116,11 @@ function link (options, event) {
 
     this.emit('ready');
 
+    var output = (this._o || this);
+
     // end event
     if (event.e) {
-        node.on('end', function () {
+        output.on('end', function () {
             instance.inst.flow(event.e, options).end();
 
             // end also error event
@@ -128,7 +133,7 @@ function link (options, event) {
 
     // error events
     if (event.r) {
-        node.on('error', function (err) {
+        output.on('error', function (err) {
             node._re = node._re || instance.inst.flow(event.r, options);
             node._re.write(err);
         });
@@ -138,9 +143,8 @@ function link (options, event) {
         node.on('error', console.error.bind(console));
     }
 
-    var output = (this._o || this);
     output.on('data', this.emit.bind(this, 'data'));
-    output.on('end', this.emit.bind(this, 'end'));
+    output.on('end', end.bind(node.scope, node, '(from output)'));
 
 //console.log('Uncork input:', node.name, '..output exists?', (this._o ? 'Yes' : 'No'));
 

--- a/lib/node.js
+++ b/lib/node.js
@@ -12,11 +12,8 @@ module.exports = function (scope, key) {
 
     node.write = node._i.write.bind(node._i);
     node.end = function (chunk) {
-		delete scope.streams[node.name];
 
-        if (node.ended) {
-            return;
-        }
+		delete scope.streams[node.name];
 
 		// wait until node is ready, so no bufferted chunks are missed
         if (!node.ready) {
@@ -25,11 +22,6 @@ module.exports = function (scope, key) {
 		    node._i.end(chunk);
         }
 	};
-
-    node.on('end', function () {
-        //console.log('Flow.node.end(event):', node.name, node.ready, node.ended);
-        delete scope.streams[node.name];
-    });
 
 	// TODO find a better solution for substring that are closed (duplex, writable)
     node.on('unpipe', node.end.bind(node));
@@ -55,18 +47,6 @@ function pipe (dest) {
     return dest;
 }
 
-function end (node, origin) {
-
-    if (node.ended) {
-		return;
-    }
-
-    //console.log('Flow.node.end:', node.name, origin, node.ready, node.ended);
-
-    node.ended = 1;
-	node.emit('end');
-}
-
 function join (output) {
     if (output) {
 
@@ -82,17 +62,14 @@ function join (output) {
 
         // pipe and set node output if ouput is duplex
         if ((output.readable && output.writable) || (output._i && output.join && output.link)) {
-//console.log((output.name || 'Seq of ' + this.name), 'Duplex pipe to', (this._o ? '"output"' : '"input"'), 'and set as output.');
             this._o = this._o ? this._o.pipe(output) : this.pipe(output);
 
         // overwrite output if sequence is readonly
         } else if (output.readable && !output.writable) {
-//console.log('Set Readable as output.');
             this._o = output;
 
         // pipe output to custom writable
         } else if (output.writable && !output.readable) {
-//console.log('Writable pipe to', (this._o ? '"output"' : '"input"'));
             (this._o || this).pipe(output);
         }
     }
@@ -104,33 +81,26 @@ function link (options, event) {
 
     node.ready = true;
 
-//console.log('\nEvent:', node.name);
-
     // pipe event transform streams
     event.s.forEach(function (sequence, index) {
         switch (sequence[0]) {
             case 0:
-
-//console.log(index + ' Data:');
-//sequence[1].forEach(function (handler) {
-//    console.log('    - handler:', handler[0][2]);
-//});
-
-                node.join(Sequence(options, sequence[1]));
+                sequence[1].forEach(function (handler) {
+                    node.join(Sequence(options, handler));
+                });
                 break;
+
             case 1:
-//console.log(index + ' Stream:');
                 options._ = sequence[1][1] || {};
                 node.join(sequence[1][0][0].call(sequence[1][0][1], options, node._o || node._i));
                 break;
+
             case 2:
-//console.log(index + ' Flow:', sequence[1][0][1]);
                 options._ = sequence[1][1] || {};
                 node.join(sequence[1][0][0].flow(sequence[1][0][1], options));
                 break;
         }
     });
-//console.log('\n');
 
     this.emit('ready');
 
@@ -162,8 +132,9 @@ function link (options, event) {
     }
 
     output.on('data', this.emit.bind(this, 'data'));
-	output.on('end', end.bind(node.scope, node, '(from output)'));
+	output.once('end', function () {
+		delete node.scope.streams[node.name];
+		node.emit('end');
+	});
     this._i.uncork();
-
-//console.log('Uncork input:', node.name, '..output exists?', (this._o ? 'Yes' : 'No'));
 }

--- a/lib/sequence.js
+++ b/lib/sequence.js
@@ -5,7 +5,6 @@ module.exports = function Sequence (options, sequence) {
     options = options || {};
 
     var seq = Stream.Transform({
-        called: 0,
         objectMode: typeof options.objectMode === 'undefined' ? true : !!options.objectMode,
         transform: function (chunk, enc, next) {
 
@@ -20,6 +19,8 @@ module.exports = function Sequence (options, sequence) {
             sequence[0][1].call(sequence[0][0], options, chunk, next, this, enc);
         }
     });
+
+    seq.called = 0;
 
     return seq;
 };

--- a/lib/sequence.js
+++ b/lib/sequence.js
@@ -9,18 +9,16 @@ module.exports = function Sequence (options, sequence) {
         transform: function (chunk, enc, next) {
 
             // once handler
-            if (this.called > 0 && sequence[3]) {
+            if (this.called && sequence[3]) {
                 return next(null, chunk);
             }
 
-            ++this.called;
+            this.called = true;
 
             options._ = sequence[1];
             sequence[0][1].call(sequence[0][0], options, chunk, next, this, enc);
         }
     });
-
-    seq.called = 0;
 
     return seq;
 };

--- a/lib/sequence.js
+++ b/lib/sequence.js
@@ -5,60 +5,19 @@ module.exports = function Sequence (options, sequence) {
     options = options || {};
 
     var seq = Stream.Transform({
+        called: 0,
         objectMode: typeof options.objectMode === 'undefined' ? true : !!options.objectMode,
         transform: function (chunk, enc, next) {
 
-            var pos = -1;
-            var handler;
-            var runSeq = function (err, data) {
-
-                // emit error
-                if (err instanceof Error || (err && data === undefined)) {
-                    err._ = chunk;
-                    return next(err);
-                }
-
-                // get handler
-                if (!(handler = sequence[++pos])) {
-
-                    // just push to next stream,
-                    // if there are no more handlers
-                    if (err && data === true) {
-                        return seq.push(err);
-                    }
-
-                    return next(null, data);
-                }
-
-                // reset handler position on push
-                if (err && data === true) {
-                    --pos;
-                    data = err;
-                }
-
-                // call next data handler
-                if (!handler[4]) {
-
-                    // once handler
-                    if (handler[3]) {
-                        handler[4] = true;
-                    }
-
-                    // attach flow composition options to caller arguments
-                    options._ = handler[1];
-
-                    // call data handler
-                    handler[0][1].call(handler[0][0], options, data, runSeq);
-                } else {
-                    runSeq(null, data);
-                }
-            };
-
-            if (!sequence) {
+            // once handler
+            if (this.called > 0 && sequence[3]) {
                 return next(null, chunk);
             }
 
-            runSeq(null, chunk);
+            ++this.called;
+
+            options._ = sequence[1];
+            sequence[0][1].call(sequence[0][0], options, chunk, next, this, enc);
         }
     });
 

